cirq-core/cirq/devices/thermal_noise_model.py

```diff
@@ -0,0 +1,251 @@
+    """
+
+    def __init__(
+        self,
+        qubit_dims: Dict['cirq.Qid', int],
+        gate_durations_ns: Dict[type, float],
+        heat_rate_GHz: Union[float, Dict['cirq.Qid', float], None] = None,
+        cool_rate_GHz: Union[float, Dict['cirq.Qid', float], None] = None,
+        dephase_rate_GHz: Union[float, Dict['cirq.Qid', float], None] = None,
+        require_physical_tag: bool = True,
+        skip_measurements: bool = True,
+    ):
+        """Construct a ThermalNoiseModel data object.
+
+        Required Args:
+            qubit_dims: Dimension for all qubits in the system.
+                        Currently only supports dimension=2 (qubits, not qudits)
+        Optional Args:
+            heat_rate_GHz: single number (units GHz) specifying heating rate,
+                        either per qubit, or global value for all.
+                        Given a rate gh, the Lindblad op will be sqrt(gh)*a^dag
+                        (where a is annihilation),
+                        so that the heating Lindbldian is
+                        gh(a^dag • a - 0.5{a*a^dag, •}).
+            cool_rate_GHz: single number (units GHz) specifying cooling rate,
+                        either per qubit, or global value for all.
+                        Given a rate gc, the Lindblad op will be sqrt(gc)*a
+                        so that the cooling Lindbldian is
+                        gc(a • a^dag - 0.5{n, •})
+                        This number is equivalent to 1/T1.
+            dephase_rate_GHz: single number (units GHz) specifying dephasing
+                        rate, either per qubit, or global value for all.
+                        Given a rate gd, Lindblad op will be sqrt(2*gd)*n where
+                        n = a^dag * a, so that the dephasing Lindbldian is
+                        2 * gd * (n • n - 0.5{n^2, •}).
+                        This number is equivalent to 1/Tphi.
+            require_physical_tag: whether to only apply noise to operations
+                        tagged with PHYSICAL_GATE_TAG.
+            skip_measurements: whether to skip applying noise to measurements.
+
+        Returns:
+            The ThermalNoiseModel with specified parameters.
+        """
+        qubits = set(qubit_dims)
+        rate_dict = {}
+
+        def _as_rate_dict(
+            rate_or_dict: Optional[Union[float, Dict['cirq.Qid', float]]]
+        ) -> Dict['cirq.Qid', float]:
+            # Convert float or None input into dictionary form. Make sure no
+            # qubits are missing from dictionary input.
+            if rate_or_dict is None:
+                return {qb: 0.0 for qb in qubits}
+            elif isinstance(rate_or_dict, dict):
+                out = rate_or_dict.copy()
+                for qb in qubits:
+                    if qb not in rate_or_dict:
+                        out[qb] = 0.0
+                return out
+            else:
+                return {qb: rate_or_dict for qb in qubits}
```

Comment:
Can we pull this out to module level and clean things up a bit ?

Reply:
Done. There might be a more concise way to express this than what I've done, but mypy is liable to complain if we condense this too much further.